### PR TITLE
dao-6

### PR DESCRIPTION
**Pull request by entertainment-veks**

### CODE REVIEW

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+import taxiservice.lib.Dao;
+import taxiservice.model.Car;
+import taxiservice.model.Driver;
+import taxiservice.model.Manufacturer;
+import taxiservice.util.ConnectionUtil;
+
+@Dao
+public class CarDaoJdbcImpl implements CarDao {
+    public static final String CARS_TAB_NAME = "cars";
+    public static final String CARS_DRIVERS_TAB_NAME = "cars_drivers";
```
Comment (Fedich09) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:25]:
```suggestion
    public static final String CARS_DRIVERS_TAB_NAME = "cars_drivers";
```
I not sure, but I think this is bad practice

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                car.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't create " + car, e);
+        }
+
+        query = "INSERT INTO " + CARS_DRIVERS_TAB_NAME + " (car_id, diver_id) VALUES (?, ?)";
```
Comment (Fedich09) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:44]: better do this in separate method, and create new connection this is bad practice, you can use same connection and use separate method in end of try block, in this way you use just one connection

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                result = newCar(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get by id " + id, e);
+        }
+
+        if (result != null) {
+            result.setDrivers(getDrivers(result.getId()));
```
Comment (Fedich09) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:78]: you can put this if/else block in try/catch and use the same connection in  method getDrivers()
```suggestion
            result.setDrivers(getDrivers(result.getId()));
```

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                result.add(newCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get all", e);
+        }
+        if (result != null) {
+            for (Car car : result) {
+                car.setDrivers(getDrivers(car.getId()));
```
Comment (Fedich09) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:101]: same here
```suggestion
                car.setDrivers(getDrivers(car.getId()));
```

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setString(1, car.getModel());
+            statement.setLong(2, car.getManufacturer().getId());
+            statement.setLong(3, car.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't update " + car, e);
+        }
+
+        query = "DELETE FROM " + CARS_DRIVERS_TAB_NAME
```
Comment (Fedich09) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:123]: create separate method for delete and use same connection, create just new PreparedStatement 
```suggestion
        query = "DELETE FROM " + CARS_DRIVERS_TAB_NAME
```

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                + " WHERE car_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, car.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't update " + car, e);
+        }
+
+        query = "INSERT INTO " + CARS_DRIVERS_TAB_NAME
```
Comment (Fedich09) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:133]:
```suggestion
        query = "INSERT INTO " + CARS_DRIVERS_TAB_NAME
```
create separate method for insert

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                result.add(newCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get all by driver", e);
+        }
+        if (result != null) {
+            for (Car car : result) {
+                car.setDrivers(getDrivers(car.getId()));
```
Comment (Fedich09) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:186]:
```suggestion
                car.setDrivers(getDrivers(car.getId()));
```
put this method inside try/catch and use one connection please

**Review by Fedich09 — CHANGES_REQUESTED** [review]

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java (hunk 1 above)
Comment (dmytro-hryhoruk) [reply src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:25]: agree,  let's write query as it is.

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java (hunk 2 above)
Comment (dmytro-hryhoruk) [reply src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:44]: also agree, let's fix this

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java (hunk 3 above)
Comment (dmytro-hryhoruk) [reply src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:78]: also agree)

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            resultManufacturer.setId(resultSet.getObject("manufacturer_id", Long.class));
+
+            Car result = new Car(
+                    resultSet.getObject("model", String.class),
+                    resultManufacturer);
+            result.setId(resultSet.getObject("car_id", Long.class));
+            return result;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't create car", e);
+        }
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:225]: let's catch these exceptions on upper lever(the places where you are calling this method)

File: src/main/java/taxiservice/dao/jdbc/DriverDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't create " + driver, e);
+        }
+        return driver;
+    }
+
+    @Override
+    public Optional<Driver> get(Long id) {
+        Driver result = new Driver(null, null);
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxiservice/dao/jdbc/DriverDaoJdbcImpl.java:42]: let's create and return Driver inside try catch

File: src/main/java/taxiservice/dao/jdbc/DriverDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        String query = "SELECT * FROM " + DRIVERS_TAB_NAME
+                + " WHERE driver_id = ? AND deleted = false";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                result.setName(resultSet.getObject("name", String.class));
+                result.setLicenceNumber(resultSet.getObject("licence_number", String.class));
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxiservice/dao/jdbc/DriverDaoJdbcImpl.java:51]: you can use newDriver() method here

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        for (Driver driver : car.getDrivers()) {
+            try (Connection connection = ConnectionUtil.getConnection();
+                     PreparedStatement statement = connection
+                             .prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {
+                statement.setLong(1, car.getId());
+                statement.setLong(2, driver.getId());
+                statement.executeUpdate();
+            } catch (SQLException e) {
+                throw new DataProcessingException("Can't create " + car, e);
+            }
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:49]: you are using this logic in update method too. Let's move it to private method

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+            return Optional.ofNullable(result);
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get by id " + id, e);
+        }
+    }
+
+    @Override
+    public List<Car> getAll() {
+        List<Car> result = new ArrayList<>();
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:79]:
```suggestion
        List<Car> cars = new ArrayList<>();
```

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java (hunk 8 above)
Comment (dmytro-hryhoruk) [reply src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:215]: ?

File: src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        query = "DELETE FROM cars_drivers WHERE car_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, car.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't update " + car, e);
+        }
```
Comment (dmytro-hryhoruk) [inline src/main/java/taxiservice/dao/jdbc/CarDaoJdbcImpl.java:123]: let's use one connection for update and delete operations(insert should be in the private method, but you can pass connection in this method too)

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**